controllers/metal3remediation_controller.go

```diff
@@ -0,0 +1,219 @@
+	host, _, err := remediationMgr.GetUnhealthyHost(ctx)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "unable to find a host for unhealthy machine")
+	}
+
+	// If user has set bmh.Spec.Online to false or annotated the host with any poweroff annotations
+	// do not try to remediate the host
+	if !remediationMgr.OnlineStatus(host) || remediationMgr.HasRebootAnnotation(host) {
+		r.Log.Info("Host is powered off, unable to remediatiate")
+	}
+
+	// If the Metal3Remediation doesn't have finalizer, add it.
+	// This is for the future work, if finalizers are needed
+	//remediationMgr.SetFinalizer()
+
+	remediationType := remediationMgr.GetRemediationType()
+
+	if remediationType == capm3.RebootRemediationStrategy {
+		// If no phase set, default to pending
+		if remediationMgr.GetRemediationPhase() == "" {
+			remediationMgr.SetRemediationPhase(capm3.PhaseRunning)
+		}
+
+		switch remediationMgr.GetRemediationPhase() {
+		case capm3.PhaseRunning:
+			// host is not rebooted yet
+			if remediationMgr.GetLastRemediatedTime() == nil {
+				r.Log.Info("Rebooting the host")
+				err := remediationMgr.SetAnnotation(ctx, "reboot")
+				if err != nil {
+					return ctrl.Result{}, errors.Wrap(err, "error setting reboot annotation")
+				}
+				now := metav1.Now()
+				remediationMgr.SetLastRemediationTime(&now)
+				remediationMgr.IncreaseRetryCount()
+			}
+
+			if remediationMgr.RetryLimitIsSet() && !remediationMgr.HasReachRetryLimit() {
+				okToRemediate, nextRemediation := remediationMgr.TimeToRemediate(remediationMgr.GetTimeout().Duration)
+
+				if okToRemediate {
+					err := remediationMgr.SetAnnotation(ctx, "reboot")
+					if err != nil {
+						return ctrl.Result{}, errors.Wrapf(err, "error setting reboot annotation")
+					}
+					now := metav1.Now()
+					remediationMgr.SetLastRemediationTime(&now)
+					remediationMgr.IncreaseRetryCount()
+				}
+
+				if nextRemediation > 0 {
+					// Not yet time to remediatate, requeue
+					return ctrl.Result{RequeueAfter: nextRemediation}, nil
+				}
+
+			} else {
+				remediationMgr.SetRemediationPhase(capm3.PhaseWaiting)
+			}
+		case capm3.PhaseWaiting:
+			okToStop, nextCheck := remediationMgr.TimeToRemediate(remediationMgr.GetTimeout().Duration)
+
+			if okToStop {
+				remediationMgr.SetRemediationPhase(capm3.PhaseDeleting)
+				// When machine is still unhealthy after remediation, setting of OwnerRemediatedCondition
+				// moves control to CAPI machine controller. The owning controller will do
+				// preflight checks and handles the Machine deletion
+				err = remediationMgr.SetOwnerRemediatedCondition(ctx)
+				if err != nil {
+					return ctrl.Result{}, errors.Wrapf(err, "error setting condition")
+				}
+
+				// Remediation failed set unhealthy annotation on BMH
+				// This prevents BMH to be selected as a host.
+				err = remediationMgr.SetAnnotation(ctx, "unhealthy")
+				if err != nil {
+					return ctrl.Result{}, errors.Wrapf(err, "error setting unhealthy annotation")
```

Comment:
missing log

Reply:
I would leave logging out from reconcile loop right now. But we may need to thing later, how we would like to do logging in the future. Since we are talking about remediation it would be good to have solid logging, but we might need to thing this over at some point.